producer/writer/writer.go

```diff
@@ -0,0 +1,258 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package writer
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/m3db/m3cluster/services"
+	"github.com/m3db/m3msg/producer"
+	"github.com/m3db/m3msg/topic"
+	xerrors "github.com/m3db/m3x/errors"
+	"github.com/m3db/m3x/log"
+	"github.com/m3db/m3x/watch"
+
+	"github.com/uber-go/tally"
+)
+
+var (
+	errWriterClosed = errors.New("writer is closed")
+)
+
+type writerMetrics struct {
+	topicUpdateSuccess  tally.Counter
+	topicUpdateError    tally.Counter
+	invalidTopicUpdate  tally.Counter
+	invalidShard        tally.Counter
+	numConsumerServices tally.Gauge
+}
+
+func newWriterMetrics(scope tally.Scope) writerMetrics {
+	return writerMetrics{
+		topicUpdateSuccess:  scope.Counter("topic-update-success"),
+		topicUpdateError:    scope.Counter("topic-update-error"),
+		invalidTopicUpdate:  scope.Counter("invalid-topic"),
+		invalidShard:        scope.Counter("invalid-shard"),
+		numConsumerServices: scope.Gauge("num-consumer-services"),
+	}
+}
+
+type writer struct {
+	sync.RWMutex
+
+	topic  string
+	ts     topic.Service
+	mPool  messagePool
+	opts   Options
+	logger log.Logger
+
+	value                  watch.Value
+	initType               initType
+	numShards              uint32
+	consumerServiceWriters map[string]consumerServiceWriter
+	filterRegistry         map[string]producer.FilterFunc
+	isClosed               bool
+	m                      writerMetrics
+
+	processFn watch.ProcessFn
+}
+
+// NewWriter creates a new writer.
+func NewWriter(opts Options) producer.Writer {
+	w := &writer{
+		topic:                  opts.TopicName(),
+		ts:                     opts.TopicService(),
+		mPool:                  newMessagePool(opts.MessagePoolOptions()),
+		opts:                   opts,
+		logger:                 opts.InstrumentOptions().Logger(),
+		initType:               failOnError,
```

Comment:
Does this need to be a field? Can it be a const?

Reply:
It's a field as it will be updated to `allowInitValueError` after the initial topic has been processed